lucille-core/src/main/java/com/kmwllc/lucille/util/FileContentFetcher.java

```diff
@@ -0,0 +1,187 @@
+package com.kmwllc.lucille.util;
+
+import com.kmwllc.lucille.connector.storageclient.StorageClient;
+import com.kmwllc.lucille.core.ConnectorException;
+import com.typesafe.config.Config;
+import java.io.BufferedReader;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.URI;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Responsible for working with StorageClients to repeatedly fetch the contents of String paths, determining the appropriate
+ * storage provider.
+ */
+public class FileContentFetcher {
+  private static final Logger log = LoggerFactory.getLogger(FileContentFetcher.class);
+
+  private final Map<String, StorageClient> availableClients;
+
+  /**
+   * Creates a FileContentFetcher that will fetch files from the classpath, the local file system, and any cloud file systems
+   * (Azure, S3, Google) which have the necessary options provided in the given options map. Be sure to call startup and shutdown
+   * as appropriate.
+   */
+  public FileContentFetcher(Config cloudOptions) {
+    this.availableClients = StorageClient.createClients(cloudOptions);
+  }
+
+  /**
+   * Initialize the storage clients associated with this FileContentFetcher. Calls shutdown on all clients and throws an Exception
+   * if an error occurs while initializing any of the StorageClients.
+   */
+  public void startup() throws IOException {
+    for (StorageClient client : availableClients.values()) {
+      try {
+        client.init();
+      } catch (ConnectorException e) {
+        shutdown();
+
+        throw new IOException("Unable to initialize StorageClient.", e);
+      }
+    }
+  }
+
+  /**
+   * Shutdown each of the storage clients associated with this FileContentFetcher. If an error occurs, a
+   * warning is logged.
+   */
+  public void shutdown() {
+    for (StorageClient client : availableClients.values()) {
+      try {
+        client.shutdown();
+      } catch (IOException e) {
+        log.warn("Error shutting down StorageClient", e);
+      }
+    }
+  }
+
+  /**
+   * Attempts to return an InputStream for the file at the given path. If the path begins with "classpath:" the prefix will be removed
+   * and the file will be read from the classpath. Otherwise, if the path is a URI, a corresponding storage client will be used;
+   * if not, the local file system will be used. Will not return a null InputStream - instead, an exception will be thrown.
+   */
+  public InputStream getInputStream(String path) throws IOException {
```

Comment:
need to validate that startup() has been called and shutdown() hasn't been called